[Add-ins] Add internet headers article (#743)
OfficeDev/outlook-dev-docs: docs/add-ins/internet-headers.md

---
title: Get and set internet headers
description: How to get and set internet headers on a message in an Outlook add-in.
ms.topic: article
ms.date: 11/01/2019
localization_priority: Normal
---

# Get and set internet headers

## Background

> It has been a common requirement in Outlook add-ins development to store custom properties associated with an add-in at different levels. At present, you can store custom properties at the item or mailbox level.

Reviewer: ```suggestion
A common requirement in Outlook add-ins development is to store custom properties associated with an add-in at different levels. At present, custom properties are stored at the item or mailbox level.
```

  }
}

// Remove custom internet headers.
function removeSelectedCustomHeaders() {
  Office.context.mailbox.item.internetHeaders.removeAsync(
    ["x-best-vegetable", "x-nonexistent-header"],
    removeCallback);
}

function removeCallback(asyncResult) {
  if (asyncResult.status === Office.AsyncResultStatus.Succeeded) {
    console.log("Successfully removed selected headers");
  } else {
    console.log("Error removing selected headers: " + JSON.stringify(asyncResult.error));
  }
}

setCustomHeaders();
getSelectedCustomHeaders();
removeSelectedCustomHeaders();
getSelectedCustomHeaders();

/* Sample output:
Successfully set headers
Selected headers: {"x-best-vegetable":"spinach","x-preferred-fruit":"orange","x-preferred-vegetable":"broccoli"}
Successfully removed selected headers
Selected headers: {"x-preferred-fruit":"orange","x-preferred-vegetable":"broccoli"}
*/
```

## Get internet headers while reading a message

You can call [item.getAllInternetHeadersAsync](/javascript/api/outlook/office.messageread#getallinternetheadersasync-options--callback-) to get internet headers on the current message in Read mode.

> ### Example

Reviewer: ```suggestion
### Get sender preferences from current MIME headers example
```

> Building on the example from the previous section, the following example shows how you can get the sender's preferences from the current email's MIME headers.

Reviewer: ```suggestion
Building on the example from the previous section, the following example shows how to get the sender's preferences from the current email's MIME headers.
```

```js
Office.context.mailbox.item.getAllInternetHeadersAsync(getCallback);

function getCallback(asyncResult) {
  if (asyncResult.status === Office.AsyncResultStatus.Succeeded) {
    console.log("Sender's preferred fruit: " + asyncResult.value.match(/x-preferred-fruit:.*/gim)[0].slice(19));
    console.log("Sender's preferred vegetable: " + asyncResult.value.match(/x-preferred-vegetable:.*/gim)[0].slice(23));
  } else {
    console.log("Error getting preferences from header: " + JSON.stringify(asyncResult.error));
  }
}

/* Sample output:
Sender's preferred fruit: orange
Sender's preferred vegetable: broccoli
*/
```

> [!IMPORTANT]

> > This sample should work for simple cases. For more complex information retrieval (e.g., multi-instance headers or folded values as described in [RFC 2822](https://tools.ietf.org/html/rfc2822)), you should use an appropriate MIME-parsing library.

Reviewer: ```suggestion
> This sample works for simple cases. For more complex information retrieval (e.g., multi-instance headers or folded values as described in [RFC 2822](https://tools.ietf.org/html/rfc2822)), try using an appropriate MIME-parsing library.
```

>

> > Also, make sure to handle a nonexistent header gracefully.

Reviewer: How? Please describe what "gracefully" means by providing an example how to handle a non-existent header, or rewrite.
```suggestion
> Also, make sure to handle a nonexistent header by including an event handler loop.
```
(I just made that up, not sure how you'd handle it, but "gracefully" is too ambiguous and doesn't localize. )
